pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,337 @@
+        declares its required resources using ``qml.register_resources``.
+
+    Args:
+        resources (dict or Callable): a dictionary mapping unique operators within the given
+            ``qfunc`` to their number of occurrences therein. If a function is provided instead
+            of a static dictionary, a dictionary must be returned from the function. For more
+            information, consult the Quantum Functions as Decomposition Rules section below.
+        qfunc (Callable): the quantum function that implements the decomposition. If ``None``,
+            returns a decorator for acting on a function.
+
+    Returns:
+        DecompositionRule:
+            a data structure that represents a decomposition rule, which contains a PennyLane
+            quantum function representing the decomposition, and its resource function.
+
+
+    **Example**
+
+    This function can be used as a decorator to bind a quantum function to its required resources
+    so that it can be used as a decomposition rule within the new graph-based decomposition system.
+
+    .. code-block:: python
+
+        import pennylane as qml
+
+        qml.decompositions.enable_graph()
+
+        @qml.register_resources({qml.H: 2, qml.CZ: 1})
+        def my_cnot(wires):
+            qml.H(wires=wires[1])
+            qml.CZ(wires=wires)
+            qml.H(wires=wires[1])
+
+        @partial(qml.transforms.decompose, fixed_decomps={qml.CNOT: my_cnot})
+        @qml.qnode(qml.device("default.qubit"))
+        def circuit():
+            qml.CNOT(wires=[0, 1])
+            return qml.state()
+
+
+    >>> print(qml.draw(circuit, level="device")())
+    0: ────╭●────┤  State
+    1: ──H─╰Z──H─┤  State
+
+    Alternatively, the decomposition rule can be created in-line:
+
+    >>> my_cnot = qml.register_resources({qml.H: 2, qml.CZ: 1}, my_cnot)
+
+    .. details::
+        :title: Quantum Functions as Decomposition Rules
+
+        Quantum functions representing decomposition rules within the new decomposition system
+        are expected to take ``(*op.parameters, op.wires, **op.hyperparameters)`` as arguments,
+        where ``op`` is an instance of the operator type that the decomposition is for.
+
+    .. details::
+        :title: Operators with Dynamic Resource Requirements
+
+        In many cases, the resource requirement of an operator's decomposition is not static; some
+        operators have properties that directly affect the resource estimate of its decompositions,
+        i.e., the types of gates that exists in the decomposition and their number of occurrences.
+        For example, the number of gates in the decomposition for ``qml.MultiRZ`` changes based
+        on the number of wires it acts on.
+
+        For each operator class, the set of parameters that affects the type of gates and their
+        number of occurrences in its decompositions is given by the ``resource_keys`` attribute:
```

Comment:
I think it reads a little better this way

```suggestion

        For each operator class, the set of parameters that affects the type of gates and their
        number of occurrences in its decompositions is given by the ``resource_keys`` attribute:
        For example, the number of gates in the decomposition for ``qml.MultiRZ`` changes based
        on the number of wires it acts on, in contrast to the decomposition for ``qml.CNOT``: